### PR TITLE
[JENKINS-40807] Fix Findbugs errors

### PR DESCRIPTION
**Pull request by varyvol**
[JENKINS-40807](https://issues.jenkins-ci.org/browse/JENKINS-40807)

After [JENKINS-40806](https://issues.jenkins-ci.org/browse/JENKINS-40806), Findbugs is run as part of the build. However, such builds does not fail if any error arise.

This change revert that behaviour and fixes/suppresses already present errors.

@reviewbybees @yacaovsnc

### CODE REVIEW

**Comment by varyvol** [comment]
@yacaovsnc did you have the opportunity to review these changes?

**Comment by yacaovsnc** [comment]
+@DavidStaheli

**Comment by jpricketMSFT** [comment]
Hi @varyvol, I am taking over from yacaovsnc. All of the changes seem reasonable. Unfortunately, I pushed some of my own changes before see this. So, I probably introduced some additional errors. I plan on merging your changes and then fixes my errors so we can get down to zero again. 
Thanks for this contribution!

**Comment by varyvol** [comment]
@jpricketMSFT Thank you!

File: tfs/src/main/java/hudson/plugins/tfs/ChangeSetReader.java [hunk 1]
```diff
@@ -25,6 +26,7 @@
 public class ChangeSetReader extends ChangeLogParser {
 
     @Override
+    @SuppressFBWarnings(value = "DM_DEFAULT_ENCODING", justification = "Better mot modify charset in case it might raise errors")
```
Comment (oleg-nenashev) [inline tfs/src/main/java/hudson/plugins/tfs/ChangeSetReader.java]: Would be better to explicitly use the System default then

File: tfs/src/main/java/hudson/plugins/tfs/TeamCollectResultsPostBuildAction.java [hunk 2]
```diff
@@ -51,6 +52,7 @@ public void setRequestedResults(final List<TeamRequestedResult> requestedResults
         this.requestedResults = requestedResults;
     }
 
+    @SuppressFBWarnings(value = "RV_RETURN_VALUE_IGNORED_BAD_PRACTICE", justification = "No matter the result of mkdirs")
```
Comment (oleg-nenashev) [inline tfs/src/main/java/hudson/plugins/tfs/TeamCollectResultsPostBuildAction.java]: Why? This is a correct issue. Ideally a Java 7 `Files` API should be used instead

File: tfs/src/main/java/hudson/plugins/tfs/TeamFoundationServerScm.java [hunk 3]
```diff
@@ -391,7 +391,7 @@ public boolean pollChanges(AbstractProject hudsonProject, Launcher launcher, Fil
     @Override
     public boolean processWorkspaceBeforeDeletion(AbstractProject<?, ?> project, FilePath workspace, Node node) throws IOException, InterruptedException {
         Run<?,?> lastRun = project.getLastBuild();
```
Comment (oleg-nenashev) [inline tfs/src/main/java/hudson/plugins/tfs/TeamFoundationServerScm.java]: You can explicitly use `AbstractBuild` here, there is no need to cast the type later

File: tfs/src/main/java/hudson/plugins/tfs/TeamPullRequestMergedDetailsAction.java [hunk 4]
```diff
@@ -24,7 +24,7 @@
     private static final long serialVersionUID = 1L;
     private static final String URL_NAME = "team-pullRequestMergedDetails";
 
-    public GitPullRequestEx gitPullRequest;
+    public transient GitPullRequestEx gitPullRequest;
```
Comment (oleg-nenashev) [inline tfs/src/main/java/hudson/plugins/tfs/TeamPullRequestMergedDetailsAction.java]: 🐛 It is going to break the event persistence on the disk, because the field won't be stored anymore.
And this action is persisted from what I see

File: tfs/src/main/java/hudson/plugins/tfs/model/GitCodePushedEventArgs.java [hunk 5]
```diff
@@ -2,10 +2,11 @@
 
 import org.eclipse.jgit.transport.URIish;
 
+import java.io.Serializable;
 import java.net.URI;
 import java.net.URISyntaxException;
 
-public class GitCodePushedEventArgs {
+public class GitCodePushedEventArgs implements Serializable {
```
Comment (oleg-nenashev) [inline tfs/src/main/java/hudson/plugins/tfs/model/GitCodePushedEventArgs.java]: It breaks the compatibility, because the class is not final.
If you make it serializable, also define the `serialVersionID`

File: tfs/src/main/java/hudson/plugins/tfs/model/ManualCredentialsConfigurer.java [hunk 6]
```diff
@@ -3,10 +3,12 @@
 import com.cloudbees.plugins.credentials.CredentialsScope;
 import com.cloudbees.plugins.credentials.common.StandardUsernamePasswordCredentials;
 import com.cloudbees.plugins.credentials.impl.UsernamePasswordCredentialsImpl;
+import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
 import hudson.Extension;
 import hudson.util.Secret;
 import org.kohsuke.stapler.DataBoundConstructor;
 
+@SuppressFBWarnings(value = "SE_TRANSIENT_FIELD_NOT_RESTORED", justification = "Maintain compatibility")
```
Comment (oleg-nenashev) [inline tfs/src/main/java/hudson/plugins/tfs/model/ManualCredentialsConfigurer.java]: There should have been readResolve() then

File: tfs/src/main/java/hudson/plugins/tfs/util/BuildWorkspaceConfigurationRetriever.java [hunk 7]
```diff
@@ -45,17 +45,23 @@ public BuildWorkspaceConfiguration getLatestForNode(Node needleNode, Run<?,?> la
     
     public static class BuildWorkspaceConfiguration extends WorkspaceConfiguration {
         private static final long serialVersionUID = 1L;
-        private final AbstractBuild<?, ?> build;
+        private final transient AbstractBuild<?, ?> build;
```
Comment (oleg-nenashev) [inline tfs/src/main/java/hudson/plugins/tfs/util/BuildWorkspaceConfigurationRetriever.java]: I am not sure what is the use-case for this class, but it may also break persistency on the disk 🐜

File: tfs/src/main/java/hudson/plugins/tfs/TeamPullRequestMergedDetailsAction.java (hunk 4 above)
Comment (varyvol) [inline tfs/src/main/java/hudson/plugins/tfs/TeamPullRequestMergedDetailsAction.java]: GitPullRequestEx is not serializable so in case someone was serializing it, it would have already broken.

File: tfs/src/main/java/hudson/plugins/tfs/model/GitCodePushedEventArgs.java (hunk 5 above)
Comment (varyvol) [inline tfs/src/main/java/hudson/plugins/tfs/model/GitCodePushedEventArgs.java]: Why does adding an interface break the compatibility? My understanding is that you can add them, being the removal the problematic thing.

https://docs.oracle.com/javase/specs/jls/se7/html/jls-13.html

File: tfs/src/main/java/hudson/plugins/tfs/model/ManualCredentialsConfigurer.java (hunk 6 above)
Comment (varyvol) [inline tfs/src/main/java/hudson/plugins/tfs/model/ManualCredentialsConfigurer.java]: Resolving to what? It's not something that can be predicted I think.

Also there are tests that are specifically checking the serialization generates an empty CredentialsConfigurer.

File: tfs/src/main/java/hudson/plugins/tfs/util/BuildWorkspaceConfigurationRetriever.java (hunk 7 above)
Comment (varyvol) [inline tfs/src/main/java/hudson/plugins/tfs/util/BuildWorkspaceConfigurationRetriever.java]: AbstractBuild is not serializable so this could have never been serialized.

**Comment by varyvol** [comment]
@oleg-nenashev I have answered some of your comments. I will create another PR with some fixes based on others.

File: tfs/src/main/java/hudson/plugins/tfs/TeamPullRequestMergedDetailsAction.java (hunk 4 above)
Comment (oleg-nenashev) [inline tfs/src/main/java/hudson/plugins/tfs/TeamPullRequestMergedDetailsAction.java]: "Serializable" is not required for persisting the data on the disk. My concern is that my making the class formally serializable you may break the data persistency

File: tfs/src/main/java/hudson/plugins/tfs/model/GitCodePushedEventArgs.java (hunk 5 above)
Comment (oleg-nenashev) [inline tfs/src/main/java/hudson/plugins/tfs/model/GitCodePushedEventArgs.java]: Serializable is a special case. If you add this interface, all child classes must become serializable as well. And there is no guarantee they really can support it.

Fromt the **binary compatibility** standpoint it is compatible, of course

File: tfs/src/main/java/hudson/plugins/tfs/util/BuildWorkspaceConfigurationRetriever.java (hunk 7 above)
Comment (oleg-nenashev) [inline tfs/src/main/java/hudson/plugins/tfs/util/BuildWorkspaceConfigurationRetriever.java]: Serialization is not required for persistency via XStream. That engine is different.
Persisting builds likely ends up with a relative link in XML, e.g. "../..". By making the field transient you are just removing it from XML at all

File: tfs/src/main/java/hudson/plugins/tfs/util/BuildWorkspaceConfigurationRetriever.java [hunk 8]
```diff
@@ -45,17 +45,23 @@ public BuildWorkspaceConfiguration getLatestForNode(Node needleNode, Run<?,?> la
...
         public BuildWorkspaceConfiguration(WorkspaceConfiguration configuration, AbstractBuild<?, ?> build) {
             super(configuration);
             this.build = build;
         }
+
         public void save() throws IOException {
             if (!workspaceExists()) {
                 build.getAction(WorkspaceConfiguration.class).setWorkspaceWasRemoved();
             }
             build.save();
```
Comment (oleg-nenashev) [inline tfs/src/main/java/hudson/plugins/tfs/util/BuildWorkspaceConfigurationRetriever.java]: risk of NPE after the change?

File: tfs/src/main/java/hudson/plugins/tfs/model/GitCodePushedEventArgs.java (hunk 5 above)
Comment (jglick) [inline tfs/src/main/java/hudson/plugins/tfs/model/GitCodePushedEventArgs.java]: Well you have to consider whether this is a plausible situation where someone might have subclassed it to begin with.

File: tfs/src/main/java/hudson/plugins/tfs/util/BuildWorkspaceConfigurationRetriever.java (hunk 7 above)
Comment (jglick) [inline tfs/src/main/java/hudson/plugins/tfs/util/BuildWorkspaceConfigurationRetriever.java]: Actually due to lazy loading issues there are cases where including a non`transient` field of type `Run` will cause serious errors, so it is mandatory to make this `transient`.

Comment (oleg-nenashev) [inline tfs/src/main/java/hudson/plugins/tfs/util/BuildWorkspaceConfigurationRetriever.java]: I agree the persisted Build instance is a bad thing in any case. But the one who solves that has to ensure the plugin's logic does not suffer. Rejuice methods or whatever. It is not enough to just make the field transient
